[WPE][GTK] REGRESSION(279912@main): fast/canvas/canvas-filter-repaint-rect.html fails

shallawa opened the pull request:
#### 5d58ef776121e63743ef363b09b6158adb97dbf0
<pre>
[WPE][GTK] REGRESSION(279912@main): fast/canvas/canvas-filter-repaint-rect.html fails
<a href="https://bugs.webkit.org/show_bug.cgi?id=275353">https://bugs.webkit.org/show_bug.cgi?id=275353</a>
<a href="https://rdar.apple.com/129719858">rdar://129719858</a>

Reviewed by Kimmo Kinnunen.

Before 279912@main, existingDrawingContext() was returning the destination context
of the canvas if it exists.

After 279912@main, existingDrawingContext() returns drawingContext() which may
return the FilterTargetSwitcher context. This causes a problem when this function
is called from RenderHTMLCanvas::requiresLayer() via isAccelerated().

Usually the canvas is backed by an accelerated ImageBuffer. But the filter target
switcher is backed by un-accelerated ImageBuffer. So returning the RenderingMode
of the target switcher will give the wrong answer to requiresLayer().

The fix is to restore the old behavior of drawingContext() back by returning the
canvas destination context and creating it if it does not exist. Also add
effectiveDrawingContext() which returns the FilterTargetSwitcher if it is active.
Otherwise it returns drawingContext().

* LayoutTests/fast/canvas/canvas-filter-bounding-rect.html:
* LayoutTests/platform/glib/TestExpectations:
* Source/WebCore/html/canvas/CanvasFilterTargetSwitcher.cpp:
(WebCore::CanvasFilterTargetSwitcher::create):
(WebCore::CanvasFilterTargetSwitcher::~CanvasFilterTargetSwitcher):
* Source/WebCore/html/canvas/CanvasRenderingContext2D.cpp:
(WebCore::CanvasRenderingContext2D::createFilter const):
(WebCore::CanvasRenderingContext2D::drawFocusIfNeededInternal):
* Source/WebCore/html/canvas/CanvasRenderingContext2DBase.cpp:
(WebCore::CanvasRenderingContext2DBase::setStrokeStyle):
(WebCore::CanvasRenderingContext2DBase::setFillStyle):
(WebCore::CanvasRenderingContext2DBase::setLineWidth):
(WebCore::CanvasRenderingContext2DBase::setLineCap):
(WebCore::CanvasRenderingContext2DBase::setLineJoin):
(WebCore::CanvasRenderingContext2DBase::setMiterLimit):
(WebCore::CanvasRenderingContext2DBase::applyLineDash const):
(WebCore::CanvasRenderingContext2DBase::setGlobalAlpha):
(WebCore::CanvasRenderingContext2DBase::setGlobalCompositeOperation):
(WebCore::CanvasRenderingContext2DBase::scale):
(WebCore::CanvasRenderingContext2DBase::rotate):
(WebCore::CanvasRenderingContext2DBase::translate):
(WebCore::CanvasRenderingContext2DBase::transform):
(WebCore::CanvasRenderingContext2DBase::setTransform):
(WebCore::CanvasRenderingContext2DBase::resetTransform):
(WebCore::CanvasRenderingContext2DBase::fillInternal):
(WebCore::CanvasRenderingContext2DBase::strokeInternal):
(WebCore::CanvasRenderingContext2DBase::clipInternal):
(WebCore::CanvasRenderingContext2DBase::beginCompositeLayer):
(WebCore::CanvasRenderingContext2DBase::endCompositeLayer):
(WebCore::CanvasRenderingContext2DBase::isPointInPathInternal):
(WebCore::CanvasRenderingContext2DBase::isPointInStrokeInternal):
(WebCore::CanvasRenderingContext2DBase::clearRect):
(WebCore::CanvasRenderingContext2DBase::fillRect):
(WebCore::CanvasRenderingContext2DBase::strokeRect):
(WebCore::CanvasRenderingContext2DBase::applyShadow):
(WebCore::CanvasRenderingContext2DBase::drawImage):
(WebCore::CanvasRenderingContext2DBase::clearCanvas):
(WebCore::CanvasRenderingContext2DBase::compositeBuffer):
(WebCore::CanvasRenderingContext2DBase::fullCanvasCompositedDrawImage):
(WebCore::CanvasRenderingContext2DBase::didDraw):
(WebCore::CanvasRenderingContext2DBase::drawingContext const):
(WebCore::CanvasRenderingContext2DBase::effectiveDrawingContext const):
(WebCore::CanvasRenderingContext2DBase::setImageSmoothingQuality):
(WebCore::CanvasRenderingContext2DBase::setImageSmoothingEnabled):
(WebCore::CanvasRenderingContext2DBase::canDrawText):
(WebCore::CanvasRenderingContext2DBase::drawTextUnchecked):
* Source/WebCore/html/canvas/CanvasRenderingContext2DBase.h:
* Source/WebCore/html/canvas/PaintRenderingContext2D.h:

Canonical link: <a href="https://commits.webkit.org/280091@main">https://commits.webkit.org/280091@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/cb4ab28573998aa4041d4cc90b4dad6ffbb5ed0a

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/55734 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/35057 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/8201 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/58718 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/59/builds/6165 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/57860 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/42679 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/6363 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/44884 "Passed tests") | [✅ 🧪 wincairo-tests](https://ews-build.webkit.org/#/builders/60/builds/4241 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/57763 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/32953 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/48042 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/26016 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/29740 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/5371 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/4308 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/51712 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/5638 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/60309 "Built successfully") | 
| | [⏳ 🛠 vision ](https://ews-build.webkit.org/#/builders/visionOS-1-Build-EWS "Waiting in queue, processing has not started yet") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/62/builds/5768 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/52311 "Passed tests") | 
| | [⏳ 🛠 vision-sim ](https://ews-build.webkit.org/#/builders/visionOS-1-Simulator-Build-EWS "Waiting in queue, processing has not started yet") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/48112 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/51811 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/12344 "Built successfully and passed tests") | [⏳ 🧪 vision-wk2 ](https://ews-build.webkit.org/#/builders/visionOS-1-Simulator-WK2-Tests-EWS "Waiting in queue, processing has not started yet") | | | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/30888 "Built successfully") | | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/31973 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/33054 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/31720 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

kkinnunen-apple commented on Source/WebCore/html/canvas/CanvasRenderingContext2DBase.cpp:
```diff
@@ -2409,7 +2409,9 @@ GraphicsContext* CanvasRenderingContext2DBase::existingDrawingContext() const
     if (!canvasBase().hasCreatedImageBuffer())
         return nullptr;
-    return drawingContext();
+    if (auto* buffer = canvasBase().buffer())
```
I think this is a bit misleading. The expectation is really that:
- drawingContext() will create the drawing buffer and return the drawing context that can draw to the drawing buffer. 
- existingDrawingContext() will return what drawingContext() would return, except null if it's not already created.



Could we instead:
```
bool CanvasRenderingContext2DBase::isAccelerated() const
{
    if (!canvasBase().hasCreatedImageBuffer())
        return false;
   auto* buffer = canvasBase().buffer();
   if (!buffer)
      return false;
   return buffer.renderingMode() == RenderingMode:::Accelerated;
}
```


Alternatively  existingDrawingContext or drawingContext could be renamed?
- drawingContext -> effectiveDrawingContext() ?

shallawa replied:
- I resorted the original behavior of `drawingContext()` back.
- So `existingDrawingContext()` is the same as `drawingContext()` without ensuring the `ImageBuffer` is created.
- I added `effectiveDrawingContext()` which returns the context of the FilterContextSwitcher if it active. Otherwise it returns `drawingContext()`. 
- I replaced all the calls to `drawingContext()` which do actual drawing by `effectiveDrawingContext()`.

kkinnunen-apple reviewed (approved):
Thanks for fixing! Use your judgement, some suggestions in the review

shallawa committed:
[WPE][GTK] REGRESSION(279912@main): fast/canvas/canvas-filter-repaint-rect.html fails
https://bugs.webkit.org/show_bug.cgi?id=275353
rdar://129719858

Reviewed by Kimmo Kinnunen.

Before 279912@main, existingDrawingContext() was returning the destination context
of the canvas if it exists.

After 279912@main, existingDrawingContext() returns drawingContext() which may
return the FilterTargetSwitcher context. This causes a problem when this function
is called from RenderHTMLCanvas::requiresLayer() via isAccelerated().

Usually the canvas is backed by an accelerated ImageBuffer. But the filter target
switcher is backed by un-accelerated ImageBuffer. So returning the RenderingMode
of the target switcher will give the wrong answer to requiresLayer().

The fix is to restore the old behavior of drawingContext() back by returning the
canvas destination context and creating it if it does not exist. Also add
effectiveDrawingContext() which returns the FilterTargetSwitcher if it is active.
Otherwise it returns drawingContext().

* LayoutTests/fast/canvas/canvas-filter-bounding-rect.html:
* LayoutTests/platform/glib/TestExpectations:
* Source/WebCore/html/canvas/CanvasFilterTargetSwitcher.cpp:
(WebCore::CanvasFilterTargetSwitcher::create):
(WebCore::CanvasFilterTargetSwitcher::~CanvasFilterTargetSwitcher):
* Source/WebCore/html/canvas/CanvasRenderingContext2D.cpp:
(WebCore::CanvasRenderingContext2D::createFilter const):
(WebCore::CanvasRenderingContext2D::drawFocusIfNeededInternal):
* Source/WebCore/html/canvas/CanvasRenderingContext2DBase.cpp:
(WebCore::CanvasRenderingContext2DBase::setStrokeStyle):
(WebCore::CanvasRenderingContext2DBase::setFillStyle):
(WebCore::CanvasRenderingContext2DBase::setLineWidth):
(WebCore::CanvasRenderingContext2DBase::setLineCap):
(WebCore::CanvasRenderingContext2DBase::setLineJoin):
(WebCore::CanvasRenderingContext2DBase::setMiterLimit):
(WebCore::CanvasRenderingContext2DBase::applyLineDash const):
(WebCore::CanvasRenderingContext2DBase::setGlobalAlpha):
(WebCore::CanvasRenderingContext2DBase::setGlobalCompositeOperation):
(WebCore::CanvasRenderingContext2DBase::scale):
(WebCore::CanvasRenderingContext2DBase::rotate):
(WebCore::CanvasRenderingContext2DBase::translate):
(WebCore::CanvasRenderingContext2DBase::transform):
(WebCore::CanvasRenderingContext2DBase::setTransform):
(WebCore::CanvasRenderingContext2DBase::resetTransform):
(WebCore::CanvasRenderingContext2DBase::fillInternal):
(WebCore::CanvasRenderingContext2DBase::strokeInternal):
(WebCore::CanvasRenderingContext2DBase::clipInternal):
(WebCore::CanvasRenderingContext2DBase::beginCompositeLayer):
(WebCore::CanvasRenderingContext2DBase::endCompositeLayer):
(WebCore::CanvasRenderingContext2DBase::isPointInPathInternal):
(WebCore::CanvasRenderingContext2DBase::isPointInStrokeInternal):
(WebCore::CanvasRenderingContext2DBase::clearRect):
(WebCore::CanvasRenderingContext2DBase::fillRect):
(WebCore::CanvasRenderingContext2DBase::strokeRect):
(WebCore::CanvasRenderingContext2DBase::applyShadow):
(WebCore::CanvasRenderingContext2DBase::drawImage):
(WebCore::CanvasRenderingContext2DBase::clearCanvas):
(WebCore::CanvasRenderingContext2DBase::compositeBuffer):
(WebCore::CanvasRenderingContext2DBase::fullCanvasCompositedDrawImage):
(WebCore::CanvasRenderingContext2DBase::didDraw):
(WebCore::CanvasRenderingContext2DBase::drawingContext const):
(WebCore::CanvasRenderingContext2DBase::effectiveDrawingContext const):
(WebCore::CanvasRenderingContext2DBase::setImageSmoothingQuality):
(WebCore::CanvasRenderingContext2DBase::setImageSmoothingEnabled):
(WebCore::CanvasRenderingContext2DBase::canDrawText):
(WebCore::CanvasRenderingContext2DBase::drawTextUnchecked):
* Source/WebCore/html/canvas/CanvasRenderingContext2DBase.h:
* Source/WebCore/html/canvas/PaintRenderingContext2D.h:

Canonical link: https://commits.webkit.org/280091@main
